### PR TITLE
Add new conv APIs to check if it uses pre-compiled kernels in DB

### PR DESCRIPTION
**Pull request by kyeonghwanryu**
I added three new APIs for convolution. They checks and return as a boolean if the kernel builds happen for the conv problem.
They work basically same with miopenFindConvolution[Direction]Algorithm, except they don't try to compile the kernels.
- miopenCheckConvolutionForwardUsePreCompiledKernel
- miopenCheckConvolutionBackwardDataUsePreCompiledKernel
- miopenCheckConvolutionBackwardWeightsUsePreCompiledKernel

I know that their names are too long. If any of the reviewers has better idea, I'm open to it.

### Context
We are having a problem with unusually long kernel build time with convolution.
As far as I know, this issue is recognized and fixed for MI300, but not for legacy GPUs. (https://github.com/ROCm/MIOpen/issues/398)
Moreh is currently using MI250 mainly and we need to bypass this issue. That's why we want to add those three APIs.

### CODE REVIEW

**Comment by atamazov** [comment]
@kyeonghwanryu Hello,
> We are having a problem with unusually long kernel build time with convolution.

Do you have a list of the kernels that take too long to build?

Do you use the regular build of MIOpen (release, with developer functions switched off)?

Do you have the precompiled kernels package installed?

**Comment by kyeonghwanryu** [comment]
@atamazov 

> Do you have a list of the kernels that take too long to build?

I met this issue about a year ago, and unfortunately it seems I didn't record the list of kernels or problems. But I could reproduce this issue by running a CNN model like ResNet50 then, so if you need those I can run it and make a list.
And I can say some of them definitely took over 20 seconds.


> Do you use the regular build of MIOpen (release, with developer functions switched off)?

We used the regular build, 5.1.0 and 5.3.3. User DB was offed for several reasons from multi-node use cases.


> Do you have the precompiled kernels package installed?

Yes.

**Comment by atamazov** [comment]
@kyeonghwanryu We appreciate the efforts you made to resolve this issue yourself. However, introducing new API calls seems like an overkill.

> But I could reproduce this issue by running a CNN model like ResNet50 then, so if you need those I can run it and make a list.
And I can say some of them definitely took over 20 seconds.

> We used the regular build, 5.1.0 and 5.3.3. 

Old releases may have bugs related to the problem you are observing. Can you please re-run the model with the most recent ROCm release. There should be no kernel builds at all provided that you are using normal release (non-DEV) build, clean environment, and precompiled kernel package installed.

**If the problem persists, then we have a bug that must be fixed**. In such a case, please provide us with the list of problematic configs (if necessary, I can provide you with the environment settings that can help to collect this information).

> User DB was offed for several reasons from multi-node use cases.

This should not affect the performance of the well-known networks (like ResNet50), but I recommend keeping it enabled. Starting from 6.0 at least, MIOpen has feature that automatically moves user databases from shared $HOME to $TMP (#2092).

**Comment by kyeonghwanryu** [comment]
> If the problem persists, then we have a bug that must be fixed.

Sorry for the late response.
I can agree with you if this issue can be handled with high priority.
 I'll check this issue with ROCm 6.0.1 and update to you. Thank you.

**Comment by atamazov** [comment]
@kyeonghwanryu 
> I can agree with you if this issue can be handled with high priority.

Sure. If this blocks your work, then we can handle this with https://github.com/ROCm/MIOpen/labels/urgency_blocker.

/cc @junliume

File: src/include/miopen/binary_cache.hpp [hunk 1]
```diff
@@ -63,6 +63,10 @@ void SaveBinary(const std::string& hsaco,
                 const std::string& args);
 #endif
 
+bool HasPreCompiledBinary(const TargetProperties& target,
+                          const size_t num_cu,
```
Comment (junliume) [inline src/include/miopen/binary_cache.hpp]: @kyeonghwanryu we have a HIP tidy issue:
```
[2024-02-29T01:37:58.009Z] /home/jenkins/workspace/n_impl_conv_kernel_build_checker/src/include/miopen/binary_cache.hpp:67:27: error: parameter 'num_cu' is const-qualified in the function declaration; const-qualification of parameters only has an effect in function definitions [readability-avoid-const-params-in-decls,-warnings-as-errors]
[2024-02-29T01:37:58.009Z]                           const size_t num_cu,
[2024-02-29T01:37:58.009Z]                           ^~~~~~
```

File: src/convolution_api.cpp [hunk 2]
```diff
@@ -1379,3 +1379,162 @@ extern "C" miopenStatus_t miopenGetConvolutionAttribute(miopenConvolutionDescrip
...
+                        w,
+                        convDesc,
+                        yDesc,
+                        y,
+                        exhaustiveSearch,
+                        ignoreAsmBuild,
+                        usePreCompiledKernel);
+
+    if(miopen::deref(convDesc).mode == miopenTranspose)
+        return miopen::try_([&] {
```
Comment (junliume) [inline src/convolution_api.cpp:1410]: @kyeonghwanryu readability tidy issue:
```
src/convolution_api.cpp:1408:56: error: statement should be inside braces [google-readability-braces-around-statements,hicpp-braces-around-statements,readability-braces-around-statements,-warnings-as-errors]
```

**Comment by atamazov** [comment]
@junliume AFAIU this PR does not require reviewing. The reason of the issue is a bug in 5.3.3 (at least) which either already fixed or will be fixed with high urgency. We are awaiting feedback from @kyeonghwanryu to continue, see https://github.com/ROCm/MIOpen/pull/2771#issuecomment-1971481767.

**Comment by kyeonghwanryu** [comment]
@atamazov 
I tried to reproduce the issue(>20s) with ROCm 6.0.2. It seems it's not happening anymore with the newest version.
I think it's better to close this PR and reopen it if the issue appears again.

And for sharing, I found the cases that took >1s for miopenFindConvolution[Direction]Algorithm.
I'm not sure it's related to the issue I've been talking about.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
</head>

<body link="#467886" vlink="#96607D">


Config | Algorithm | Time (ms)
-- | -- | --
64-56-56-1x1-64-56-56-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,583
256-56-56-1x1-128-56-56-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,416
128-28-28-1x1-512-28-28-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,538
256-56-56-1x1-512-28-28-16-0x0-2x2-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,447
512-28-28-1x1-256-28-28-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,152
512-28-28-1x1-256-28-28-16-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoImplicitGEMM | 2,858
256-56-56-1x1-128-56-56-16-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoImplicitGEMM | 2,816
256-56-56-1x1-512-28-28-1-0x0-2x2-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,321
1024-14-14-1x1-256-14-14-1-0x0-1x1-1x1-0-NCHW-FP32-B | miopenConvolutionBwdDataAlgoImplicitGEMM | 1,176
64-56-56-1x1-256-56-56-1-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoDirect | 1,002
64-56-56-1x1-64-56-56-1-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoDirect | 1,012
64-56-56-1x1-256-56-56-8-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,787
256-56-56-1x1-128-56-56-8-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,672
3-224-224-11x11-64-55-55-16-2x2-4x4-1x1-1-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 4,076
256-13-13-3x3-256-13-13-1-1x1-1x1-1x1-1-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 4,375
3-224-224-11x11-64-55-55-8-2x2-4x4-1x1-1-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,818



</body>

</html>

**Comment by kyeonghwanryu** [comment]
I tried to reproduce this issue(>20s) with the ROCm 6.0.2 but it seems it's not happening anymore with the newest version. I think it's better to close this PR and reopen it if it appears again.

And for sharing, I found some cases that took >1s for miopenFindConvolution[Direction]Algorithm. I'm not sure it's related to this issue.
 
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/maloja/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/maloja/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
</head>

<body link="#467886" vlink="#96607D">


Config | Algorithm | Time (ms)
-- | -- | --
64-56-56-1x1-64-56-56-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,583
256-56-56-1x1-128-56-56-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,416
128-28-28-1x1-512-28-28-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,538
256-56-56-1x1-512-28-28-16-0x0-2x2-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,447
512-28-28-1x1-256-28-28-16-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,152
512-28-28-1x1-256-28-28-16-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoImplicitGEMM | 2,858
256-56-56-1x1-128-56-56-16-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoImplicitGEMM | 2,816
256-56-56-1x1-512-28-28-1-0x0-2x2-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,321
1024-14-14-1x1-256-14-14-1-0x0-1x1-1x1-0-NCHW-FP32-B | miopenConvolutionBwdDataAlgoImplicitGEMM | 1,176
64-56-56-1x1-256-56-56-1-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoDirect | 1,002
64-56-56-1x1-64-56-56-1-0x0-1x1-1x1-0-NCHW-FP32-W | miopenConvolutionBwdWeightsAlgoDirect | 1,012
64-56-56-1x1-256-56-56-8-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,787
256-56-56-1x1-128-56-56-8-0x0-1x1-1x1-0-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,672
3-224-224-11x11-64-55-55-16-2x2-4x4-1x1-1-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 4,076
256-13-13-3x3-256-13-13-1-1x1-1x1-1x1-1-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 4,375
3-224-224-11x11-64-55-55-8-2x2-4x4-1x1-1-NCHW-FP32-F | miopenConvolutionFwdAlgoImplicitGEMM | 2,818



</body>

</html>

**Pull request closed without merging**